db/cache_db.go

```diff
@@ -0,0 +1,157 @@
+package db
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+)
+
+// If value is nil but deleted is false,
+// it means the parent doesn't have the key.
+// (No need to delete upon Commit())
+type cDBValue struct {
+	value   []byte
+	deleted bool
+}
+
+// CacheDB wraps an in-memory cache around an underlying DB.
+type CacheDB struct {
+	mtx    sync.Mutex
+	cache  map[string]cDBValue
+	parent DB
+}
+
+// NOTE: Needed by MultiStore.CacheWrap().
+var _ atomicSetDeleter = (*CacheDB)(nil)
+
+func NewCacheDB(parent DB) *CacheDB {
+	db := &CacheDB{
+		cache:  make(map[string]cDBValue),
+		parent: parent,
+	}
+	return db
+}
+
+func (db *CacheDB) Get(key []byte) []byte {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	value, ok := db.cache[string(key)]
+	if !ok {
+		data := db.parent.Get(key)
+		value = cDBValue{value: data}
+		db.cache[string(key)] = value
+	}
+	return value.value
+}
+
+func (db *CacheDB) Set(key []byte, value []byte) {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
```

Comment:
Now here I am being a bit painful about this, but I believe we can squeeze even better perhaps by avoiding defers whenever we can, as they are ~2X slower than non-defers, please see https://gist.github.com/odeke-em/f8fb3a8695b33dd06a700fc3473f3e62.
Could we please make all these:
```go
db.mtx.Lock()
db.SetNoLock(key, value)
db.mtx.Unlock()
```

Reply:
I insist that we keep the defers until we find that it's a bottleneck... it's defensive programming that IMO is well worth it.